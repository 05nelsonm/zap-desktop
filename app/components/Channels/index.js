export ChannelCapacity from './ChannelCapacity'
export ChannelMoreButton from './ChannelMoreButton'
export ChannelStatus from './ChannelStatus'
export ChannelSummaryList from './ChannelSummaryList'
export ChannelSummaryListItem from './ChannelSummaryListItem'
