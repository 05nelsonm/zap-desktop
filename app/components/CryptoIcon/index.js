import CryptoIcon from './CryptoIcon'

export default CryptoIcon
