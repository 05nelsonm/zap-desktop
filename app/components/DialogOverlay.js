import styled from 'styled-components'
import { Flex } from 'rebass'

const Overlay = styled(Flex)`
  position: absolute;
  z-index: 99999;
  background-color: rgba(255, 255, 255, 0.25);
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
`

export default Overlay
