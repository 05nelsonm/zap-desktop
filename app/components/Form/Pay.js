import React, { Component } from 'react'
import PropTypes from 'prop-types'

import Isvg from 'react-inlinesvg'
import paperPlane from 'icons/paper_plane.svg'

import { FaBolt, FaChain, FaAngleDown } from 'react-icons/lib/fa'
import LoadingBolt from 'components/LoadingBolt'
import CurrencyIcon from 'components/CurrencyIcon'

import styles from './Pay.scss'

class Pay extends Component {
  componentDidUpdate(prevProps) {
    const {
      isOnchain, isLn, payform: { payInput }, fetchInvoice
    } = this.props

    // If on-chain, focus on amount to let user know it's editable
    if (isOnchain) { this.amountInput.focus() }

    // If LN go retrieve invoice details
    if ((prevProps.payform.payInput !== payInput) && isLn) {
      fetchInvoice(payInput)
    }
  }

  render() {
    const {
      payform: { amount, payInput, showErrors },
      currency,
      crypto,

      isOnchain,
      isLn,
      currentAmount,
      usdAmount,
      inputCaption,
      showPayLoadingScreen,
      payFormIsValid: { errors, isValid },

      setPayAmount,
      onPayAmountBlur,

      setPayInput,
      onPayInputBlur,

      onPaySubmit
    } = this.props

    console.log('usdAmount: ', usdAmount)

    return (
      <div className={styles.container}>
        {showPayLoadingScreen && <LoadingBolt />}
        <header className={styles.header}>
          <Isvg src={paperPlane} />
          <h1>Make Payment</h1>
        </header>

        <div className={styles.content}>
          <section className={styles.destination}>
            <div className={styles.top}>
              <label htmlFor='destination'>Destination</label>
              <span>
              </span>
            </div>
            <div className={styles.bottom}>
              <textarea
                type='text'
                placeholder='Payment request or bitcoin address'
                value={payInput}
                onChange={event => setPayInput(event.target.value)}
                onBlur={onPayInputBlur}
                id='destination'
                rows='4'
              />
            </div>
          </section>

          <section className={styles.amount}>
            <div className={styles.top}>
              <label>Amount</label>
              <span></span>
            </div>
            <div className={styles.bottom}>
              <input
                type='number'
                min='0'
                ref={(input) => { this.amountInput = input }}
                size=''
                placeholder='0.00000000'
                value={currentAmount || ''}
                onChange={event => setPayAmount(event.target.value)}
                onBlur={onPayAmountBlur}
                id='amount'
                readOnly={isLn}
              />
              <div className={styles.currency}>
                <section className={styles.currentCurrency}>
                  <span>BTC</span><span><FaAngleDown /></span>
                </section>
                <ul>
                  <li>Bits</li>
                  <li>Satoshis</li>
                </ul>
              </div>
            </div>

            <div className={styles.usdAmount}>
              {`â‰ˆ ${usdAmount} USD`}
            </div>
          </section>

          <section className={styles.submit}>
            <div className={`${styles.button} ${isValid && styles.active}`} onClick={onPaySubmit}>Pay</div>
          </section>
        </div>
      </div>
    )
  }
}


Pay.propTypes = {
  payform: PropTypes.shape({
    amount: PropTypes.string.isRequired,
    payInput: PropTypes.string.isRequired,
    showErrors: PropTypes.object.isRequired
  }).isRequired,
  currency: PropTypes.string.isRequired,
  crypto: PropTypes.string.isRequired,

  isOnchain: PropTypes.bool.isRequired,
  isLn: PropTypes.bool.isRequired,
  currentAmount: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  inputCaption: PropTypes.string.isRequired,
  showPayLoadingScreen: PropTypes.bool.isRequired,
  payFormIsValid: PropTypes.shape({
    errors: PropTypes.object,
    isValid: PropTypes.bool
  }).isRequired,

  setPayAmount: PropTypes.func.isRequired,
  onPayAmountBlur: PropTypes.func.isRequired,
  setPayInput: PropTypes.func.isRequired,
  onPayInputBlur: PropTypes.func.isRequired,
  fetchInvoice: PropTypes.func.isRequired,

  onPaySubmit: PropTypes.func.isRequired
}

export default Pay
