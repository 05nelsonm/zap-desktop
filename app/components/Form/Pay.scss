@import '../../variables.scss';

.container {
  padding: 0 40px;
  font-family: Roboto;
}

.header {
  text-align: center;
  padding-bottom: 20px;
  color: $white;
  border-bottom: 1px solid $spaceborder;

  h1 {
    font-size: 22px;
    font-weight: 100;
    margin-top: 10px;
    letter-spacing: 1.5px;
  }
}

.content {
  margin-top: 50px;
  color: $white;

  .destination {
    margin-bottom: 25px;
  }

  .amount .bottom {
    display: flex;
    flex-direction: row;
    align-items: center;

    input {
      font-size: 40px;
      max-width: 250px;
    }
  }

  .top {
    margin-bottom: 30px;

    label {
      font-size: 14px;
    }
  }

  .bottom {
    input, textarea {
      background: transparent;
      outline: none;
      border: 0;
      color: $gold;
      -webkit-text-fill-color: $white;
      font-size: 12px;
      width: 100%;
      font-weight: 200;
    }

    input::-webkit-input-placeholder, textarea::-webkit-input-placeholder {
      text-shadow: none;
      -webkit-text-fill-color: initial;
    }
  }

  .currency {
    display: flex;
    flex-direction: row;
    align-items: center;

    .currentCurrency {
      cursor: pointer;
      transition: 0.25s all;

      &:hover {
        opacity: 0.5;
      }

      span {
        font-size: 14px;
        
        &:nth-child(1) {
          font-weight: bold;
        }
      }

    }

    ul {
      visibility: hidden;
      position: absolute;
    }
  }

  .usdAmount {
    margin-top: 20px;
  }

  .submit {
    margin-top: 50px;
    text-align: center;

    .button {
      width: 235px;
      margin: 0 auto;
      padding: 20px 10px;
      background: #31343f;
      opacity: 0.5;
      cursor: pointer;
      transition: 0.25s all;

      &.active {
        background: $gold;
        opacity: 1.0;

        &:hover {
          background: darken($gold, 5%);
        }
      }
    }
  }
}
