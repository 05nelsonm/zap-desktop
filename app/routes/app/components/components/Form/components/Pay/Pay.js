import React from 'react'
import PropTypes from 'prop-types'
import { FaBolt, FaChain } from 'react-icons/lib/fa'
import CurrencyIcon from '../../../../../../../components/CurrencyIcon'
import { btc } from '../../../../../../../utils'
import styles from './Pay.scss'

const Pay = ({
  paymentType,
  setPaymentType,
  invoiceAmount,
  onchainAmount,
  setOnchainAmount,
  payment_request,
  setPaymentRequest,
  fetchInvoice,
  payInvoice,
  currency,
  crypto,
  close
}) => {
  const payClicked = () => {
    if (payment_request.length !== 124) { return }

    payInvoice(payment_request)
    close()
  }

  const paymentRequestOnChange = payreq => {
    setPaymentRequest(payreq)
    if (payreq.length === 124) {
      setPaymentType('ln')
      fetchInvoice(payreq) 
    } else if (payreq.length === 42) {
      setPaymentType('onchain')
    } else {
      setPaymentType('')
    }
  }
  
  const calculateAmount = value => (currency === 'usd' ? btc.satoshisToUsd(value, currentTicker.price_usd) : btc.satoshisToBtc(value))

  return (
    <div className={styles.container}>
      <section className={`${styles.amountContainer} ${paymentType === 'ln' ? styles.ln : ''}`}>
        <label htmlFor='amount'>
          <CurrencyIcon currency={currency} crypto={crypto} />
        </label>
        <input
          type='text'
          size=''
          style={
            paymentType === 'ln' ?
              { width: '75%', fontSize: '85px' }
              :
              { width: `${onchainAmount.length > 1 ? (onchainAmount.length * 15) - 5 : 25}%`, fontSize: `${190 - (onchainAmount.length ** 2)}px` }
          }
          value={paymentType === 'ln' ? calculateAmount(invoiceAmount) : onchainAmount}
          onChange={event => setOnchainAmount(event.target.value)}
          id='amount'
          readOnly={paymentType === 'ln'}
        />
      </section>
      <div className={styles.inputContainer}>
        <div className={styles.info}>
          {(() => {
            switch(paymentType) {
              case 'onchain':
                return (
                  <span>{`You're about to send ${onchainAmount} on-chain which should take around 10 minutes`}</span>
                )
              case 'ln':
                return (
                  <span>{`You're about to send ${calculateAmount(invoiceAmount)} over the Lightning Network which will be instant`}</span>
                )
              default:
                return null
            }
          })()}
        </div>
        <aside className={styles.paymentIcon}>
          {(() => {
            switch(paymentType) {
              case 'onchain':
                return (
                  <i>
                    <span>on-chain</span>
                    <FaChain />
                  </i>
                )
              case 'ln':
                return (
                  <i>
                    <span>lightning network</span>
                    <FaBolt />
                  </i>
                )
              default:
                return null
            }
          })()}
        </aside>
        <section className={styles.input}>
          <input
            type='text'
            placeholder='Payment request or bitcoin address'
            value={payment_request}
            onChange={event => paymentRequestOnChange(event.target.value)}
            id='paymentRequest'
          />
        </section>
      </div>
      <section className={styles.buttonGroup}>
        <div className={styles.button} onClick={payClicked}>
          Pay
        </div>
      </section>
    </div>
  )
}

Pay.propTypes = {
  amount: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  setOnchainAmount: PropTypes.func.isRequired,
  setPaymentRequest: PropTypes.func.isRequired,
  fetchInvoice: PropTypes.func.isRequired,
  payInvoice: PropTypes.func.isRequired,
  currency: PropTypes.string.isRequired,
  crypto: PropTypes.string.isRequired,
  close: PropTypes.func.isRequired
}

export default Pay
