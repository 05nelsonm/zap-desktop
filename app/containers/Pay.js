import { connect } from 'react-redux'
import { Pay } from 'components/Pay'
import { tickerSelectors } from 'reducers/ticker'
import { setPayReq, queryRoutes } from 'reducers/pay'
import { sendCoins } from 'reducers/transaction'
import { payInvoice } from 'reducers/payment'

const mapStateToProps = state => ({
  chain: state.info.chain,
  network: state.info.network,
  cryptoName: tickerSelectors.cryptoName(state),
  channelBalance: state.balance.channelBalance,
  cryptoCurrency: state.ticker.currency,
  cryptoCurrencyTicker: tickerSelectors.currencyName(state),
  payReq: state.pay.payReq,
  onchainFees: state.pay.onchainFees,
  routes: state.pay.routes,
  walletBalance: state.balance.walletBalance
})

const mapDispatchToProps = {
  payInvoice,
  setPayReq,
  sendCoins,
  queryRoutes
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Pay)
