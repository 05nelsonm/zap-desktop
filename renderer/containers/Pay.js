import { connect } from 'react-redux'
import { Pay } from 'components/Pay'
import { tickerSelectors } from 'reducers/ticker'
import { setPayReq, queryFees, queryRoutes } from 'reducers/pay'
import { balanceSelectors } from 'reducers/balance'
import { changeFilter } from 'reducers/activity'
import { sendCoins } from 'reducers/transaction'
import { payInvoice } from 'reducers/payment'
import { closeModal } from 'reducers/modal'

const mapStateToProps = state => ({
  chain: state.info.chain,
  network: state.info.network,
  cryptoName: tickerSelectors.cryptoName(state),
  channelBalance: balanceSelectors.channelBalance(state),
  cryptoCurrency: state.ticker.currency,
  cryptoCurrencyTicker: tickerSelectors.currencyName(state),
  isQueryingFees: state.pay.isQueryingFees,
  payReq: state.pay.payReq,
  onchainFees: state.pay.onchainFees,
  routes: state.pay.routes,
  walletBalanceConfirmed: balanceSelectors.walletBalanceConfirmed(state),
})

const mapDispatchToProps = {
  changeFilter,
  closeModal,
  payInvoice,
  setPayReq,
  sendCoins,
  queryFees,
  queryRoutes,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Pay)
