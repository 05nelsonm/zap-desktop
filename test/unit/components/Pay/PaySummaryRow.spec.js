import React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'
import { PaySummaryRow } from 'components/Pay'

describe('component.Form.PaySummaryRow', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<PaySummaryRow left="left contnet" right="right content" />)
    expect(toJSON(wrapper)).toMatchSnapshot()
  })
})
