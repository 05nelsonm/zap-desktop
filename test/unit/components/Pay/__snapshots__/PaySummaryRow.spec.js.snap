// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component.Form.PaySummaryRow should render correctly 1`] = `
<styled.div
  py={2}
>
  <Styled(styled.div)
    alignItems="center"
  >
    <styled.div
      width={0.5}
    >
      <Text
        fontWeight="normal"
      >
        left contnet
      </Text>
    </styled.div>
    <styled.div
      width={0.5}
    >
      <Text
        textAlign="right"
      >
        right content
      </Text>
    </styled.div>
  </Styled(styled.div)>
</styled.div>
`;
